# Lab | Intro to Probability

**Objective**

Welcome to this Intro to Probability lab, where we explore decision-making scenarios through the lens of probability and strategic analysis. In the business world, making informed decisions is crucial, especially when faced with uncertainties. This lab focuses on scenarios where probabilistic outcomes play a significant role in shaping strategies and outcomes. Students will engage in exercises that require assessing and choosing optimal paths based on data-driven insights. The goal is to enhance your skills by applying probability concepts to solve real-world problems.

**Challenge 1**

#### Ironhack Airlines 

Often Airlines sell more tickets than they have seats available, this is called overbooking. Consider the following:
- A plane has 450 seats. 
- Based on historical data we conclude that each individual passenger has a 3% chance of missing it's flight. 

If the Ironhack Airlines routinely sells 460 tickets, what is the chance that they have a seats for all passenger?

In [1]:
from scipy.stats import binom

# Parameters
n = 460        # tickets sold
p = 0.97       # probability that a passenger shows up
capacity = 450 # number of seats

# Probability that all passengers have a seat
prob = binom.cdf(capacity, n, p)

print(f"The probability that all passengers have a seat is {prob:.4%}")


The probability that all passengers have a seat is 88.4477%


**Challenge 2**

#### Ironhack Call Center 

Suppose a customer service representative at a call center is handling customer complaints. Consider the following:
- The probability of successfully resolving a customer complaint on the first attempt is 0.3. 


What is the probability that the representative needs to make at least three attempts before successfully resolving a customer complaint?

In [2]:
from scipy.stats import geom

p = 0.3  # probability of success on each attempt

# Probability that at least 3 attempts are needed
prob = 1 - geom.cdf(2, p)

print(f"The probability that at least 3 attempts are needed is {prob:.2%}")

The probability that at least 3 attempts are needed is 49.00%


**Challenge 3**

#### Ironhack Website

Consider a scenario related to Ironhack website traffic. Where:
- our website takes on average 500 visits per hour.
- the website's server is designed to handle up to 550 vists per hour.


What is the probability of the website server being overwhelmed?

In [3]:
from scipy.stats import poisson

# Parameters
lam = 500    # average visits per hour
capacity = 550

# Probability that the server is overwhelmed
prob = 1 - poisson.cdf(capacity, lam)

print(f"The probability that the server is overwhelmed is {prob:.4%}")


The probability that the server is overwhelmed is 1.2898%


What is the probability of being overwhelmed at some point during a day? (consider 24hours)

In [4]:
# Parameters
lam = 500    # average visits per hour
capacity = 550

# Probability of being overwhelmed in a single hour
p_hour = 1 - poisson.cdf(capacity, lam)

# Probability of being overwhelmed at least once in 24 hours
p_day = 1 - (1 - p_hour) ** 24

print(f"Probability of being overwhelmed in one hour: {p_hour:.4%}")
print(f"Probability of being overwhelmed at least once in a day: {p_day:.4%}")


Probability of being overwhelmed in one hour: 1.2898%
Probability of being overwhelmed at least once in a day: 26.7704%


**Challenge 4**

#### Ironhack Helpdesk

Consider a scenario related to the time between arrivals of customers at a service desk.

On average, a customers arrives every 10minutes.

What is the probability that the next customer will arrive within the next 5 minutes?

In [5]:
import math
from scipy.stats import expon

# Parameters
mean = 10            # average arrival time in minutes
lam = 1 / mean       # rate per minute

# Probability using formula
prob_formula = 1 - math.exp(-lam * 5)

# Probability using scipy
prob_scipy = expon.cdf(5, scale=mean)

print(f"Probability (formula): {prob_formula:.4%}")
print(f"Probability (scipy): {prob_scipy:.4%}")


Probability (formula): 39.3469%
Probability (scipy): 39.3469%


If there is no customer for 15minutes, employees can that a 5minutes break.

What is the probability an employee taking a break?

In [6]:
import math
from scipy.stats import expon

mean = 10          # average arrival time in minutes
lam = 1 / mean     # rate per minute

# Probability that no customer arrives within 15 minutes
prob_formula = math.exp(-lam * 15)

# Using scipy
prob_scipy = 1 - expon.cdf(15, scale=mean)  # survival function
# or directly: expon.sf(15, scale=mean)

print(f"Probability (formula): {prob_formula:.4%}")
print(f"Probability (scipy): {prob_scipy:.4%}")


Probability (formula): 22.3130%
Probability (scipy): 22.3130%


22.31% chance an employee can take a break.

**Challenge 5**

The weights of a certain species of birds follow a normal distribution with a mean weight of 150 grams and a standard deviation of 10 grams. 

- If we randomly select a bird, what is the probability that its weight is between 140 and 160 grams?

In [7]:
from scipy.stats import norm

# Parameters
mu = 150
sigma = 10

# Probability that weight is between 140 and 160
prob = norm.cdf(160, mu, sigma) - norm.cdf(140, mu, sigma)

print(f"Probability that a bird weighs between 140g and 160g: {prob:.2%}")


Probability that a bird weighs between 140g and 160g: 68.27%


**Challenge 6**

If the lifetime (in hours) of a certain electronic component follows an exponential distribution with a mean lifetime of 50 hours, what is the probability that the component fails within the first 30 hours?

In [8]:
import math
from scipy.stats import expon

mean = 50          # average lifetime in hours
lam = 1 / mean     # rate per hour

# Probability using formula
prob_formula = 1 - math.exp(-lam * 30)

# Probability using scipy
prob_scipy = expon.cdf(30, scale=mean)

print(f"Probability (formula): {prob_formula:.2%}")
print(f"Probability (scipy): {prob_scipy:.2%}")


Probability (formula): 45.12%
Probability (scipy): 45.12%
